### PR TITLE
Implemented RFC 4178 extension 

### PR DESCRIPTION
**Pull request by tengcm2015**
reference issue #50

### CODE REVIEW

File: gssapi/raw/ext_rfc4178.pyx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        OM_uint32 *minor_status,
+        gss_cred_id_t cred_handle,
+        const gss_OID_set mechs) nogil
+
+
+def set_neg_mechs(Creds cred_handle not None, mechs not None):
+    """
+    set_neg_mechs(Creds cred_handle not None, mechs=None)
+
+    This allows callers to specify the set of security mechanisms that
```
Comment (frozencemetery) [inline gssapi/raw/ext_rfc4178.pyx]: We don't typically use the RFC text for descriptions - it tends to be overly lengthy and include a lot of details we don't care about.  See if you can make a shorter description in your own words.

File: gssapi/raw/ext_rfc4178.pyx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    may be negotiated with the credential identified by cred_handle.
+    This call is intended to support specialized callers who need to
+    restrict the set of negotiable security mechanisms from the set of
+    all security mechanisms available to the caller (based on available
+    credentials). Note that if more than one mechanism is specified in
+    mech_set, the order in which those mechanisms are specified implies a
+    relative preference.
+
+    Args:
+        cred_handle (Creds): the Credentials to query
```
Comment (frozencemetery) [inline gssapi/raw/ext_rfc4178.pyx]: Technically not a query - that would be more true of `gss_get_neg_mechs()` (which no one implements).

File: gssapi/raw/ext_rfc4178.pyx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        mechs ([MechType]): the desired set of mechanisms that
```
Comment (frozencemetery) [inline gssapi/raw/ext_rfc4178.pyx]: Maybe shorten this too

File: gssapi/raw/ext_rfc4178.pyx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            may be negotiated with the credentials
+    Returns:
+        None (If the operation completed successfully)
```
Comment (frozencemetery) [inline gssapi/raw/ext_rfc4178.pyx]: The parenthetical is unnecessary.

File: gssapi/raw/ext_rfc4178.pyx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        maj_stat = gss_set_neg_mechs(&min_stat, cred_handle.raw_creds,
+                                     desired_mechs)
+
+    cdef OM_uint32 tmp_min_stat
+    gss_release_oid_set(&tmp_min_stat, &desired_mechs)
+
+    if maj_stat == GSS_S_COMPLETE:
+        return None
+    else:
+        print("error: %d:%d", maj_stat, min_stat)
```
Comment (frozencemetery) [inline gssapi/raw/ext_rfc4178.pyx]: Leftover debugging line?

File: gssapi/tests/test_raw.py [hunk 6]
```diff
@@ -772,6 +772,34 @@ def test_sasl_names(self):
             cmp_mech.shouldnt_be_none()
             cmp_mech.should_be(mech)
 
+    @ktu.gssapi_extension_test('rfc4178', 'Negotiation Mechanism')
+    def test_set_neg_mechs(self):
```
Comment (frozencemetery) [inline gssapi/tests/test_raw.py]: While I think we want to enforce mech_set not being None (even though 4178 doesn't require it), I would like to see us test mech selection.  We should have the NTLM mechanism around - maybe something using that?  (See test_set_sec_context_option and test_set_sec_context_option_fail.)  You'll need to install gssntlmssp to check this locally.

**Comment by frozencemetery** [comment]
Extension detection doesn't seem to have worked on Heimdal because our Debian's Heimdal is too old.  (It was added this year in 735039dbdc3aa58d06afdefd214efe3f5e421244.)  However, it'll work whenever that updates - they just put it in gssapi.h.

@DirectXMan12 PTAL

**Comment by DirectXMan12** [comment]
my context for this is mostly lost to time.  Which part did you want me to take a look at?  The extension negotiation?

**Comment by frozencemetery** [comment]
I was just hoping for code review :)

**Comment by frozencemetery** [comment]
@tengcm2015 Hey, this is still waiting on you.  Any updates?

**Comment by tengcm2015** [comment]
@frozencemetery I fixed what is mentioned in the code review, but have no idea why the test is failing only on centOS

**Comment by frozencemetery** [comment]
Does it work for you locally on centos?

**Comment by tengcm2015** [comment]
Well, I guess the problem was that centOS had no ntmlssp, but since I created a credential with all_mech, it just created a credential with some other mech rather than skipping the test.
